lib/class-wp-theme-json.php

```diff
@@ -0,0 +1,953 @@
+<?php
+/**
+ * Process of structures that adhere to the theme.json schema.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Class that encapsulates the processing of
+ * structures that adhere to the theme.json spec.
+ */
+class WP_Theme_JSON {
+
+	/**
+	 * Container of data in theme.json format.
+	 *
+	 * @var array
+	 */
+	private $contexts = null;
+
+	/**
+	 * Holds block metadata extracted from block.json
+	 * to be shared among all instances so we don't
+	 * process it twice.
+	 *
+	 * @var array
+	 */
+	private static $blocks_metadata = null;
+
+	/**
+	 * Data schema of each context within a theme.json.
+	 *
+	 * Example:
+	 *
+	 * {
+	 *   'context-one': {
+	 *     'styles': {
+	 *       'color': {
+	 *         'background': 'color'
+	 *       }
+	 *     },
+	 *     'settings': {
+	 *       'color': {
+	 *         'custom': true
+	 *       }
+	 *     }
+	 *   },
+	 *   'context-two': {
+	 *     'styles': {
+	 *       'color': {
+	 *         'link': 'color'
+	 *       }
+	 *     }
+	 *   }
+	 * }
+	 */
+	const SCHEMA = array(
+		'selector' => null,
```

Comment:
Specifying the schema pointing things can contain strings like selector to null seems strange.
What if specify in the same format to specify inputs for the rest API:
Example:
```
/**
 * Get our sample schema for comments.
 */
function prefix_get_comment_schema() {
    $schema = array(
        // This tells the spec of JSON Schema we are using which is draft 4.
        '$schema'              => 'http://json-schema.org/draft-04/schema#',
        // The title property marks the identity of the resource.
        'title'                => 'comment',
        'type'                 => 'object',
        // In JSON Schema you can specify object properties in the properties attribute.
        'properties'           => array(
            'id' => array(
                'description'  => esc_html__( 'Unique identifier for the object.', 'my-textdomain' ),
                'type'         => 'integer',
                'context'      => array( 'view', 'edit', 'embed' ),
                'readonly'     => true,
            ),
            'author' => array(
                'description'  => esc_html__( 'The id of the user object, if author was a user.', 'my-textdomain' ),
                'type'         => 'integer',
            ),
            'content' => array(
                'description'  => esc_html__( 'The content for the object.', 'my-textdomain' ),
                'type'         => 'string',
            ),
        ),
    );
 
    return $schema;
}
```

Reply:
You mean, adding type validation at the leaf node level?

It crossed my mind, and I think it'd be potentially helpful if connected with the logging mechanism for theme authors to see, which is something they have suggested they'd want. It can be useful, yes, but perhaps can be added in a subsequent PR? Would love to land this one quickly to unblock some work, let other people use this, and avoid many rebases (which are going to be / already were quite painful and time-consuming).

Reply:
Yah having a schema would allow validation at the node level, and I guess developers could even use validation tools to make sure their theme.json is correct.
I guess the first version of a schema could just validate what we have currently but specifying things in schema.json would be better than "'selector' => null," as the current syntax is a little confusing, selector is not null it is a string. It seems we are inventing our own schema.
I agree it can be done as a follow up 👍

Reply:
Created https://github.com/WordPress/gutenberg/issues/26955 to track this.